src/indexer/services/evmIndexer/evmIndexer.ts

```diff
@@ -38,40 +70,138 @@ export class EvmIndexer {
         toBlock += this.pastEventsQueryInterval
       } catch (error) {
-        logger.error(`Failed to process past events because of: ${error}`)
+        logger.error(`Failed to process past events because of: ${(error as Error).message}`)
       }
     } while (fromBlock < latestBlock)
     // move to next block from the last queried range in past events
     return latestBlock + 1
   }
 
   async listenToEvents(): Promise<void> {
-    logger.info(`Starting querying current blocks for events on ${this.domain.name}`)
+    logger.info(`Starting querying blocks for events on ${this.domain.name}, domainID: ${this.domain.id}`)
     let latestBlock = await this.indexPastEvents()
-    this.provider.on("block", async (currentBlock: number) => {
-      // start at last block from past events query and move to new blocks range
-      if (latestBlock + this.currentEventsQueryInterval === currentBlock) {
-        // connect executions to deposits
-        try {
-          // fetch and decode logs
-          await this.saveDataToDb(this.domain.id, currentBlock.toString(), this.domain.name)
-          // move to next range of blocks
-          latestBlock += this.currentEventsQueryInterval
-        } catch (error) {
-          logger.error(`Failed to process current events because of: ${error}`)
+    await this.provider.on("block", (currentBlock: number): void => {
+      const handleBlockEvent = async (): Promise<void> => {
+        // start at last block from past events query and move to new blocks range
+        if (latestBlock + this.currentEventsQueryInterval === currentBlock) {
+          // connect executions to deposits
+          try {
+            // fetch and decode logs
+            const logs = await getLogs(this.provider, this.domain, latestBlock, currentBlock)
+            if (logs.length > 0) {
+              logger.debug(`Found past events on ${this.domain.name} in block range [${latestBlock}-${currentBlock}]`)
+            }
+            const decodedLogs = await decodeLogs(this.provider, this.domain, logs, this.resourceMap)
+            await this.saveDataToDb(decodedLogs, this.domain.id, currentBlock.toString(), this.domain.name)
+            // move to next range of blocks
+            latestBlock += this.currentEventsQueryInterval
+            return
+          } catch (error) {
+            logger.error(`Failed to process current events because of: ${(error as Error).message}`)
+          }
+          return undefined
         }
       }
+      void handleBlockEvent()
     })
   }
 
-  async saveDataToDb(domainID: number, latestBlock: string, domainName: string): Promise<void> {
-    logger.info(`save block on ${domainName}: ${latestBlock}`)
-    this.domainRepository.upserDomain(domainID, latestBlock, domainName)
+  async saveDataToDb(decodedLogs: DecodedLogs, domainID: number, latestBlock: string, domainName: string): Promise<void> {
+    try {
+      const transferMap = new Map<string, string>()
+
+      await Promise.all(
```

Comment:
I would prefer all this promise.alls to be functions to clean it up a bit.

Reply:
done